main/lsp/requests/definition.cc

```diff
@@ -30,17 +30,20 @@ LSPResult LSPLoop::handleTextDocumentDefinition(unique_ptr<core::GlobalState> gs
         if (!queryResponses.empty()) {
             auto resp = move(queryResponses[0]);
 
-            if (auto identResp = resp->isIdent()) {
-                for (auto &originLoc : identResp->retType.origins) {
+            if (resp->isIdent() || resp->isConstant() || resp->isLiteral()) {
+                auto retType = resp->getTypeAndOrigins();
+                for (auto &originLoc : retType.origins) {
                     addLocIfExists(*gs, result, originLoc);
                 }
             } else if (auto defResp = resp->isDefinition()) {
                 result.push_back(loc2Location(*gs, defResp->termLoc));
-            } else {
-                for (auto &component : resp->getDispatchComponents()) {
-                    if (component.method.exists() && !component.receiver->isUntyped()) {
-                        addLocIfExists(*gs, result, component.method.data(*gs)->loc());
+            } else if (auto sendResp = resp->isSend()) {
+                auto start = sendResp->dispatchResult.get();
+                while (start != nullptr) {
+                    if (start->main.method.exists() && !start->main.receiver->isUntyped()) {
+                        addLocIfExists(*gs, result, start->main.method.data(*gs)->loc());
                     }
+                    start = start->secondary.get();
                 }
```

Comment:
Is there a sensible behavior for a catch-all down here? If we were to ever add a different kind of LSP query response, then we'd just fall-through here without reporting a location, which isn't a bad thing, but it might be nice to get an explicit error that we forgot to cover a case rather than having some silent anomalous LSP behavior in that situation.

Reply:
I think this was intentional, `null` is a valid return for any query and thus IDE tries to be robust.
That said, I think the new behavior is equivalent to pre-existing one. I'll let @jvilk-stripe decide what does he want here long term, maybe in a separate PR.

Reply:
`null` or an empty vector is a valid response in ~all requests where the user points to a particular character. For example, the user might request a definition / references / hover at a location that only contains whitespace.

The code here defaults to sending an empty vector as the catch-all behavior.